plugins/thunderbolt/fu-thunderbolt-device.c

```diff
@@ -0,0 +1,645 @@
+	if (did == 0x0)
+		g_debug ("failed to get Device ID");
+
+	/* requires kernel 5.5 or later, non-fatal if not available */
+	gen = fu_thunderbolt_device_get_attr_uint16 (self, "generation", &error_gen);
+	if (gen == 0)
+		g_debug ("Unable to read generation: %s", error_gen->message);
+
+	/* determine if host controller or not */
+	if (parent_name != NULL && g_str_has_prefix (parent_name, "domain")) {
+		self->host = TRUE;
+		fu_device_add_flag (device, FWUPD_DEVICE_FLAG_INTERNAL);
+		fu_device_set_summary (device, "Unmatched performance for high-speed I/O");
+	} else {
+		tmp = fu_udev_device_get_sysfs_attr (FU_UDEV_DEVICE (self), "device_name", NULL);
+	}
+
+	/* set the controller name */
+	if (tmp == NULL) {
+		if (gen == 4)
+			tmp = "USB4 Controller";
+		else
+			tmp = "Thunderbolt Controller";
+	}
+	fu_device_set_name (device, tmp);
+
+	/* set vendor string */
+	tmp = fu_udev_device_get_sysfs_attr (FU_UDEV_DEVICE (self), "vendor_name", error);
+	if (tmp == NULL)
+		return FALSE;
+	fu_device_set_vendor (device, tmp);
+
+	/* try to read the version */
+	if (!fu_thunderbolt_device_get_version (self)) {
+		/* failed to read, for host check for safe mode */
+		if (self->host)
+			fu_thunderbolt_device_check_safe_mode (self);
+	}
+
+	if (self->safe_mode) {
+		fu_device_set_update_error (device, "Device is in safe mode");
+	} else {
+		g_autofree gchar *device_id = NULL;
+		g_autofree gchar *domain_id = NULL;
+		if (fu_thunderbolt_device_can_update (self)) {
+			g_autofree gchar *vendor_id = NULL;
+			g_autofree gchar *domain = g_path_get_basename (self->devpath);
+			gboolean is_native = FALSE;
+			/* USB4 controllers don't have a concept of legacy vs native
+			 * so don't try to read a native attribute from their NVM */
+			if (self->host && gen < 4) {
+				g_autoptr(GError) native_error = NULL;
+				if (!fu_thunderbolt_device_is_native (self,
+								      &is_native,
+								      &native_error)) {
+					g_warning ("failed to get native mode status: %s",
+						   native_error->message);
+				}
+				domain_id = g_strdup_printf ("TBT-%04x%04x%s-controller%s",
+							     (guint) vid,
+							     (guint) did,
+							     is_native ? "-native" : "",
+							     domain);
+			}
+			vendor_id = g_strdup_printf ("TBT:0x%04X", (guint) vid);
+			fu_device_set_vendor_id (device, vendor_id);
+			device_id = g_strdup_printf ("TBT-%04x%04x%s",
+						     (guint) vid,
+						     (guint) did,
+						     is_native ? "-native" : "");
+			fu_device_add_flag (device, FWUPD_DEVICE_FLAG_UPDATABLE);
+			fu_device_add_flag (device, FWUPD_DEVICE_FLAG_DUAL_IMAGE);
+		} else {
+			device_id = g_strdup ("TBT-fixed");
+			fu_device_set_update_error (device, "Missing non-active nvmem");
+		}
+		fu_device_add_instance_id (device, device_id);
+		if (domain_id != NULL)
+			fu_device_add_instance_id (device, domain_id);
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_plugin_thunderbolt_trigger_update (FuThunderboltDevice	*self,
```

Comment:
Something I'm considering here is if we should instead kick this into a delayed activation (or maybe make that behavior optional).

If we did that then potentially then thunderbolt devices could set a flag that they are usable during an update and once we reboot/shutdown apply the update.

The downside is unplugging a Thunderbolt or USB4 device without rebooting will cause firmware to not be updated.  So it is definitely a trade off.

Reply:
I think if we can apply it live we can -- rebooting is a real pain the neck for most users.